contracts/BarterMarket.sol

```diff
@@ -148,15 +148,113 @@ contract BarterWalletFactory {
     offerCount += 1;
   }
 
-  // TODO: add a function to accept an offer from another address
-  // Check that all assets have been approved for transfer
-  // Check that ETH balances are sufficient
-  // Check that the offer is still in the SENT state
-  // Send the ETH, tokens and NFTs to the target
-  // Sent the ETH, tokens and NFTs to the offerer
-  // Reset committed tokens and NFTs
-  // Change the state to ACCEPTED
-  // Emit an event of a TradeOffer being accepted
+	function acceptOffer(uint256 _offerId) public payable {
+		TradeOffer memory offer = offers[_offerId];
+		
+		require(
+			offer.target == msg.sender, 
+			"This offer was not sent to you"
+		);
+		require(
+			offer.askBundle.offeredEther == msg.value,
+			"Offer amount is not equal to the amount of ETH sent"
+		);
+		require(
+			offer.state == State.SENT,
+			"This offer is no longer available"
+		);
+		
+		// State should be set before transfer i think https://medium.com/coinmonks/common-attacks-in-solidity-and-how-to-defend-against-them-9bc3994c7c18
+		// If anything fails all state changes will be reverted anyways, kind of like a db transaction
+		offers[_offerId].state = State.ACCEPTED;
+
+		// Check that the target's assets are approved for the trade, then transfer
+		CoinBundle memory askCoins = offer.askBundle.tokens;			
+		NFTBundle memory askNfts = offer.askBundle.nfts;
+
+		for (uint256 i = 0; i < askCoins.contractAddresses.length; i++) {
+			ERC20 coinContract = ERC20(askCoins.contractAddresses[i]);
+			
+			require(
+				coinContract.balanceOf(offer.target) >= askCoins.amounts[i],
+				"Not enough tokens"
+			);
+			require(
+				coinContract.allowance(offer.target, address(this)) >=
+					askCoins.amounts[i] +
+						committedTokens[offer.target][askCoins.contractAddresses[i]],
+				"Not enough allowed tokens"
+			);
```

Comment:
I think we can get slight gas optimization by skipping the first require check if we assume that `allowance` will never return a value higher than `balanceOf`

Reply:
I'm not sure that's a safe assumption. from https://ethereum.org/en/developers/tutorials/erc20-annotated-code/ :

"This is the function that actually specifies allowances. Note that it allows an owner to specify an allowance that is higher than the owner's current balance. This is OK because the balance is checked at the time of transfer, when it could be different from the balance when the allowance is created."